src/app/clusters/thread-border-router-management-server/thread-br-mgmt-server.cpp

```diff
@@ -0,0 +1,384 @@
+    }
+
+    CHIP_ERROR err = mDelegate->GetDataset(dataset, type);
+    if (err != CHIP_NO_ERROR)
+    {
+        return err == CHIP_IM_GLOBAL_STATUS(NotFound) ? StatusIB(err).mStatus : Status::Failure;
+    }
+    return Status::Success;
+}
+
+Status ServerInstance::HandleSetActiveDatasetRequest(bool failSafeArmed,
+                                                     const Commands::SetActiveDatasetRequest::DecodableType & req)
+{
+    // The SetActiveDatasetRequest command SHALL be FailSafeArmed. Upon receiving this command, the Thread BR will set its
+    // active dataset. If the dataset is set successfully, OnActivateDatasetComplete will be called with CHIP_NO_ERROR, prompting
+    // the Thread BR to respond with a success status and disarm the FailSafe timer. If an error occurs while setting the active
+    // dataset, the Thread BR should respond with a failure status. In this case, when the FailSafe timer expires, the active
+    // dataset set by this command will be reverted. If the FailSafe timer expires before the Thread BR responds, the Thread BR will
+    // respond with a timeout status and the active dataset should also be reverted.
+    ReturnErrorCodeIf(!failSafeArmed, Status::FailsafeRequired);
+    ReturnErrorCodeIf(!mDelegate, Status::InvalidInState);
+
+    Thread::OperationalDataset activeDataset;
+    Thread::OperationalDataset currentActiveDataset;
+    uint64_t currentActiveDatasetTimestamp;
+    // If any of the parameters in the ActiveDataset is invalid, the command SHALL fail with a status code
+    // of INVALID_COMMAND.
+    ReturnErrorCodeIf(activeDataset.Init(req.activeDataset) != CHIP_NO_ERROR, Status::InvalidCommand);
+
+    // If this command is invoked when the ActiveDatasetTimestamp attribute is not null, the command SHALL
+    // fail with a status code of INVALID_IN_STATE.
+    if (mDelegate->GetDataset(currentActiveDataset, Delegate::DatasetType::kActive) == CHIP_NO_ERROR &&
+        currentActiveDataset.GetActiveTimestamp(currentActiveDatasetTimestamp) == CHIP_NO_ERROR)
+    {
+        return Status::InvalidInState;
+    }
+
+    mBreadcrumb    = req.breadcrumb;
+    CHIP_ERROR err = mDelegate->SetActiveDataset(activeDataset, this);
+    return StatusIB(err).mStatus;
+}
+
+Status ServerInstance::HandleSetPendingDatasetRequest(const Commands::SetPendingDatasetRequest::DecodableType & req)
+{
+    ReturnErrorCodeIf(!mDelegate, Status::InvalidInState);
+    bool panChangeSupported;
+    if (mDelegate->GetPanChangeSupported(panChangeSupported) != CHIP_NO_ERROR || !panChangeSupported)
+    {
+        return Status::UnsupportedCommand;
+    }
+    Thread::OperationalDataset pendingDataset;
+    // If any of the parameters in the PendingDataset is invalid, the command SHALL fail with a status code
+    // of INVALID_COMMAND.
+    ReturnErrorCodeIf(pendingDataset.Init(req.pendingDataset) != CHIP_NO_ERROR, Status::InvalidCommand);
+    CHIP_ERROR err = mDelegate->SetPendingDataset(pendingDataset);
```

Comment:
What if there is already a pending dataset on the network, and the pending dataset timestamp in the existing dataset is higher than the timestamp in this pending dataset? That will result in a silent failure where the dataset with the later timestamp will be accepted

Reply:
I believe that should be add to SPEC so that I will update the server logic according to the SPEC.

Reply:
I'm confused by this comment... is there a timestamp check here? What should be the change to the spec?

Reply:
There should be a check for a pending dataset that may already be active on the network. If there is, and the pending timestamp in that dataset is higher, then any attempt to set a pending dataset with a lower timestamp will silently fail. Whatever dataset has the highest timestamp will be selected.

Reply:
What do you mean by "set a pending dataset with a lower timestamp"? The SetPendingDatasetRequest Command does not have a timestamp field.

Reply:
It does; its a field in the operational dataset sent with the `req` arg (`req.pendingDataset`). It may be acceptable behavior that the attempt to set a pending dataset can fail in some cases i.e. there already is a pending dataset with a higher timestamp on the network waiting to become the active dataset. But I think some logging / returning a relevant error code would be useful to at least give the caller the option for a retry, something akin to `EAGAIN`. So may be a good idea to add a check for a pending dataset already on the network at this point